StanfordAHA/garnet
# Interconnect Redesign

Changes to the first interconnect:
1. Based on #146, the graph and circuit creation is completely separated. Users now interact with the graph instead of the circuit
2. Add configuration register to each CB, SB, and Tile level. Verified in `verilator`.
3. Add more tests on the interconnect tiling


CB Tests:
- Test all the possible mux connections and `sel_bits`.
SB Tests
- Test all the possible routes within a switch box based on `sel_bits`.
- Tested with different bit width, various number of tracks, and different type of switch box designs.

Tile Test
- Test all possible routes through the tile. All configured with configuration registers
  - Switch box node -> Core input
  - Core output -> switch box node
  - Tested with different number of tracks

~TODO:~
~1. Add configuration to `Interconnect` Class~
~2. Add tests to configure a PnR routing result on the interconnect~
~3. Clean up some unused code and add missing tests.~

Future Work for the next PR (by this Friday):
1. Add PnR collateral support (partially done in this PR)
2. Add pipeline registers

Discussion:
The global signal needs to be redesign given the new `Interconnect`, which already handles the configuration address and data. Another way to deal with this is to refactor out the configuration logic into a different class, yet it imposes some challenges to the unit testing: we need a way to automatically lift the selection ports from the muxs to the top level.
Update: I discussed the global signal issue with @alexcarsello, we will work together to create another pass to deal with the global signal based on the existing code we have.

EDIT:
Finished the PR. It should be good to go.
